fix: adding DNSADDR protocol

acul71 opened the pull request:
I'm working on implementing DNSADDR protocol.
Initial release.
Many test are failing. 
Work in progress....

acul71 commented:
@seetadev @arcinston Ready for Review
# DNSADDR Protocol Feature

## Overview

The DNSADDR protocol is a new feature added to the py-multiaddr-fork that enables DNS-based address resolution for multiaddrs. This allows multiaddrs to be resolved to their actual network addresses at runtime, supporting both IPv4 and IPv6 addresses.

## What is DNSADDR?

DNSADDR is a protocol that allows multiaddrs to contain DNS hostnames instead of hardcoded IP addresses. When resolved, these hostnames are looked up using DNS A (IPv4) and AAAA (IPv6) records to obtain the actual network addresses.

### Protocol Code
- **Protocol Code**: `0x38` (56 in decimal)
- **Protocol Name**: `dnsaddr`
- **Codec**: `domain`

## Implementation

### Core Components

#### 1. DNS Resolver (`multiaddr/resolvers/dns.py`)

The DNS resolver is implemented as an async resolver that supports:
- **DNS A records** (IPv4 addresses)
- **DNS AAAA records** (IPv6 addresses)
- **Recursive resolution** with configurable depth limits
- **Cancellation support** for timeout handling
- **Peer ID encapsulation** for p2p addresses

#### 2. Resolver Interface (`multiaddr/resolvers/base.py`)

Defines the base protocol for resolvers:
```python
class Resolver(Protocol):
    async def resolve(self, maddr: Multiaddr) -> list[Multiaddr]:
        """Resolve a multiaddr to its final form."""
```

### Key Features

#### Recursive Resolution
DNSADDR supports recursive resolution with a configurable maximum depth (default: 32 levels) to prevent infinite loops.

#### Cancellation Support
The resolver supports cancellation through Trio's `CancelScope`, allowing for timeout-based resolution and graceful cancellation of long-running DNS lookups.

#### Peer ID Support
When resolving DNSADDR addresses that include peer IDs, the resolver preserves the peer ID in the final resolved addresses.

#### Error Handling
- **RecursionLimitError**: When maximum recursive depth is exceeded
- **ResolutionError**: When DNS resolution fails
- **Graceful fallback**: Returns original multiaddr when resolution fails

## Usage Examples

### Basic DNSADDR Resolution

```python
from multiaddr import Multiaddr
from multiaddr.resolvers import DNSResolver

# Create a DNS resolver
resolver = DNSResolver()

# Resolve a DNSADDR multiaddr
ma = Multiaddr("/dnsaddr/example.com")
resolved = await resolver.resolve(ma)
# Result: [Multiaddr("/ip4/93.184.216.34"), Multiaddr("/ip6/2606:2800:220:1:248:1893:25c8:1946")]
```

### DNSADDR with Peer ID

```python
# DNSADDR with embedded peer ID
ma = Multiaddr("/dnsaddr/example.com/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7wjh53Qk")
resolved = await resolver.resolve(ma)
# Result: [Multiaddr("/ip4/93.184.216.34/p2p/QmYyQSo1c1Ym7orWxLYvCrM2EmxFTANf8wXmmE7wjh53Qk")]
```

### With Custom Options

```python
# Custom recursive depth and cancellation signal
import trio

signal = trio.CancelScope()
options = {
    "max_recursive_depth": 5,
    "signal": signal
}

resolved = await resolver.resolve(ma, options)
```

### Direct Multiaddr Resolution

```python
# Resolve directly on the multiaddr object
ma = Multiaddr("/dnsaddr/example.com")
resolved = await ma.resolve()
# Same result as using the resolver directly
```

## Testing

### Test Coverage

The DNSADDR feature includes comprehensive tests in `tests/test_resolvers.py`:

#### Basic Functionality Tests
- `test_resolve_dns_addr`: Tests basic DNSADDR resolution
- `test_resolve_dns_addr_with_peer_id`: Tests resolution with peer IDs
- `test_resolve_recursive_dns_addr`: Tests recursive resolution

#### Error Handling Tests
- `test_resolve_recursion_limit`: Tests recursion limit enforcement
- `test_resolve_dns_addr_error`: Tests DNS resolution failure handling

#### Edge Cases
- `test_resolve_dns_addr_with_quotes`: Tests hostname cleaning
- `test_resolve_dns_addr_with_mixed_quotes`: Tests mixed quote handling

#### Cancellation Tests
- `test_resolve_cancellation_with_error`: Tests cancellation support

### Mock Strategy

Tests use mocked DNS responses to ensure reliability:
```python
# Mock DNS resolver to return predictable results
mock_answer_a = AsyncMock()
mock_rdata_a = AsyncMock()
mock_rdata_a.address = "127.0.0.1"
mock_answer_a.__iter__.return_value = [mock_rdata_a]

with patch.object(dns_resolver._resolver, 'resolve') as mock_resolve:
    mock_resolve.side_effect = mock_resolve_side_effect
    # Test DNS resolution
```

## Configuration Options

### Resolver Options

The DNS resolver accepts the following options:

- **`max_recursive_depth`** (int): Maximum depth for recursive resolution (default: 32)
- **`signal`** (trio.CancelScope): Cancellation signal for timeout handling

### Default Settings

- **Default timeout**: 5.0 seconds
- **Maximum recursive depth**: 32 levels
- **Supported record types**: A (IPv4) and AAAA (IPv6)

## Technical Details

### DNS Record Access

The implementation uses proper type casting to access DNS record addresses:

```python
# For A records (IPv4)
address = str(cast(dns.rdtypes.IN.A.A, rdata).address)

# For AAAA records (IPv6)
address = str(cast(dns.rdtypes.IN.AAAA.AAAA, rdata).address)
```

### Type Checking Compatibility

Due to version mismatches between runtime and type stubs, the implementation includes type ignore comments:

```python
with trio.CancelScope() as cancel_scope:  # type: ignore[call-arg]
    cancel_scope.cancelled_caught = True
```

### Error Recovery

When DNS resolution fails, the resolver gracefully returns the original multiaddr:

```python
try:
    resolved = await self._resolve_dnsaddr(hostname, maddr, max_depth, signal)
    return resolved if resolved else [maddr]
except Exception as e:
    return [maddr]  # Fallback to original
```

## Benefits

1. **Dynamic Addressing**: Multiaddrs can point to hostnames that resolve to different IPs
2. **Load Balancing**: DNS can provide different IPs for load balancing
3. **Geographic Distribution**: DNS can return geographically appropriate IPs
4. **Failover**: DNS can provide backup IPs when primary addresses fail
5. **IPv6 Support**: Automatic support for both IPv4 and IPv6 addresses

## Future Enhancements

Potential improvements for the DNSADDR feature:
- Support for additional DNS record types (SRV, TXT)
- Caching of DNS resolutions
- Configurable DNS servers
- Support for DNS over HTTPS (DoH)
- Integration with other resolver types

acul71 commented:
@arcinston 
# Trio CancelScope Type Checking and Runtime Compatibility

## Problem

When using `trio.CancelScope()` in this codebase, the type checker (pyright) reports an error:

```
reportCallIssue: Argument missing for parameter "cancelled_caught"
```

However, at runtime (with Trio 0.30.0), passing `cancelled_caught` to the constructor raises:

```
TypeError: CancelScope.__init__() got an unexpected keyword argument 'cancelled_caught'
```

## Investigation

### 1. Check Trio Version

```bash
$ python -c "import trio; print(trio.__version__)"
0.30.0
```

### 2. Check CancelScope Constructor Signature

```bash
$ python -c "import trio; help(trio.CancelScope.__init__)"
Help on function __init__ in module trio:

__init__(
    self,
    *,
    relative_deadline: 'float' = inf,
    deadline: 'float' = inf,
    shield: 'bool' = False
) -> None
    Method generated by attrs for class CancelScope.
```

**Result:** No `cancelled_caught` parameter in the constructor.

### 3. Check CancelScope Attributes

```bash
$ python -c "import trio; print(dir(trio.CancelScope))"
['__annotations__', ..., 'cancel', 'cancel_called', 'cancelled_caught', ...]
```

**Result:** `cancelled_caught` is a property, not a constructor argument.

### 4. Set Property After Construction

```python
import trio
scope = trio.CancelScope()
print(scope.cancelled_caught)  # False
scope.cancelled_caught = True
print(scope.cancelled_caught)  # True
```

**Output:**
```
False
True
```

## Solution

- **Runtime:** Use `trio.CancelScope()` with no arguments, then set `cancelled_caught = True` after creation.
- **Type checker:** Add `# type: ignore[call-arg]` to suppress the error, since the type stubs expect a constructor argument that does not exist in this version.

### Example

```python
with trio.CancelScope() as cancel_scope:  # type: ignore[call-arg]
    cancel_scope.cancelled_caught = True
    ...
```

## Why is this needed?
- The type stubs for Trio may be for a newer version than the one installed, causing a mismatch.
- This approach ensures both runtime correctness and type checking pass.

pacrob commented on tests/test_resolvers.py:
```diff
@@ -0,0 +1,237 @@
+@pytest.mark.trio
+async def test_resolve_dns_addr(dns_resolver):
+    """Test resolving a DNS multiaddr."""
+    # Create mock DNS answer for A record (IPv4)
+    mock_answer_a = AsyncMock()
+    mock_rdata_a = AsyncMock()
+    mock_rdata_a.address = "127.0.0.1"
+    mock_answer_a.__iter__.return_value = [mock_rdata_a]
+
+    # Create mock DNS answer for AAAA record (IPv6) - return empty to avoid conflicts
+    mock_answer_aaaa = AsyncMock()
+    mock_answer_aaaa.__iter__.return_value = []
+
+    with patch.object(dns_resolver._resolver, 'resolve') as mock_resolve:
+        # Configure the mock to return different results based on record type
+        async def mock_resolve_side_effect(hostname, record_type):
+            if record_type == "A":
+                return mock_answer_a
+            elif record_type == "AAAA":
+                return mock_answer_aaaa
+            else:
+                raise dns.resolver.NXDOMAIN()
```
The mock code is repeated a lot. Probably worth pulling it out to a fixture.

acul71 replied:
Sure, refactored the duplicate mock DNS resolution code into a reusable fixture

pacrob commented on tests/test_resolvers.py:
```diff
@@ -0,0 +1,237 @@
+            await trio.sleep(0.1)
+            signal.cancel()
+
+            # The nursery should handle the cancellation
```
Should there be some assertion here? Or it will throw an exception if it is not handled correctly?

acul71 replied:
Yes thanks, assertion was missing. 
The test now asserts that signal.cancel_called is True, confirming that cancellation was actually triggered.

acul71 commented on multiaddr/resolvers/__init__.py:
```diff
@@ -0,0 +1,26 @@
+        Returns:
+            A list of resolved multiaddrs
+        """
+        ...
```
Sure, removing here, and importing instead

acul71 commented on tests/test_resolvers.py:
```diff
@@ -0,0 +1,237 @@
+
+        mock_resolve.side_effect = mock_resolve_side_effect
+
+        ma = Multiaddr("/dnsaddr/example.com")
+        result = await dns_resolver.resolve(ma)
+        assert len(result) == 1
+        assert result[0].protocols()[0].name == "ip4"
+        assert result[0].value_for_protocol(result[0].protocols()[0].code) == "127.0.0.1"
```
Sorry  The "mixed quotes" test now actually tests mixed quotes functionality instead of being a duplicate of the previous test

acul71 commented:
@pacrob Ops, I didn't know I could merge myself. Is this ok?

## Recent Commits

- **757aec3** — `test: assert cancellation is handled in test_resolve_cancellation_with_error`  
  Added an explicit assertion to ensure DNS resolution cancellation is handled and tested properly.

- **ce716bf** — `fix: implement actual mixed quotes testing in DNS resolver test`  
  Updated the test to actually verify mixed quotes handling in DNS resolution.

- **a7d8681** — `refactored the duplicate mock DNS resolution code into a reusable fixture`  
  Reduced code duplication in tests by introducing a fixture for mock DNS resolution.

- **b39118f** — `chore: removed repeated code class Resolver`  
  Removed duplicate `Resolver` protocol definition to ensure a single source of truth.

pacrob commented:
> @pacrob Ops, I didn't know I could merge myself. Is this ok?

Yes, it's fine. You'll want to do a squash instead of just merge in the future though, it keeps the commit history cleaner. 

Thanks for getting this in!
